File: frontend/src/api/services/hierarchical.ts
import apiClient from '../client';

export interface HierarchicalModelCreate {
  name: string;
  parent_model_id?: string;
  dataset_id: string;
  dimension_columns: string[];
  granularity_type: string;
  model_type: string;
  target_variable: string;
  date_column?: string;
  features?: Array<{
    column: string;
    enabled: boolean;
    adstock?: { enabled: boolean; decay: number; max_lag: number };
    saturation?: { enabled: boolean; type: string; half_saturation?: number };
  }>;
  inherit_constraints: boolean;
  constraint_relaxation: number;
  inherit_priors: boolean;
  prior_weight: number;
  min_observations: number;
}

export interface SubModel {
  id: string;
  dimension_values: Record<string, string>;
  status: 'pending' | 'training' | 'completed' | 'failed' | 'skipped';
  error_message?: string;
  observation_count?: number;
  r_squared?: number;
  rmse?: number;
  training_duration_seconds?: number;
  model_config_id?: string;
  created_at: string;
  updated_at: string;
}

export interface HierarchicalModel {
  id: string;
  project_id: string;
  name: string;
  parent_model_id?: string;
  dataset_id: string;
  dimension_columns: string[];
  granularity_type: string;
  model_type: string;
  target_variable: string;
  date_column?: string;
  features?: Array<Record<string, unknown>>;
  inherit_constraints: boolean;
  constraint_relaxation: number;
  inherit_priors: boolean;
  prior_weight: number;
  min_observations: number;
  status: string;
  task_id?: string;
  sub_models: SubModel[];
  created_at: string;
  updated_at: string;
}

export interface HierarchicalModelListItem {
  id: string;
  name: string;
  granularity_type: string;
  model_type: string;
  status: string;
  sub_model_count: number;
  completed_count: number;
  created_at: string;
}

export interface TrainingProgress {
  total: number;
  completed: number;
  failed: number;
  in_progress: number;
  pending: number;
}

export interface TrainingStatus {
  status: string;
  progress: TrainingProgress;
  sub_models: SubModel[];
}

export interface DimensionCombination {
  values: Record<string, string>;
  observation_count: number;
}

export interface DimensionAnalysis {
  dimension_columns: string[];
  combinations: DimensionCombination[];
  total_combinations: number;
}

export interface ResultsSummary {
  id: string;
  name: string;
  total_sub_models: number;
  completed_sub_models: number;
  avg_r_squared?: number;
  min_r_squared?: number;
  max_r_squared?: number;
  coefficient_comparisons: Array<{
    variable: string;
    national_estimate?: number;
    national_ci_lower?: number;
    national_ci_upper?: number;
    sub_model_estimates: Record<string, number>;
  }>;
}

export const hierarchicalApi = {
  list: async (projectId: string): Promise<HierarchicalModelListItem[]> => {
    const response = await apiClient.get<HierarchicalModelListItem[]>(
      `/projects/${projectId}/hierarchical-models`
    );
    return response.data;
  },

  create: async (projectId: string, data: HierarchicalModelCreate): Promise<HierarchicalModel> => {
    const response = await apiClient.post<HierarchicalModel>(
      `/projects/${projectId}/hierarchical-models`,
      data
    );
    return response.data;
  },

  get: async (projectId: string, configId: string): Promise<HierarchicalModel> => {
    const response = await apiClient.get<HierarchicalModel>(
      `/projects/${projectId}/hierarchical-models/${configId}`
    );
    return response.data;
  },

  delete: async (projectId: string, configId: string): Promise<void> => {
    await apiClient.delete(`/projects/${projectId}/hierarchical-models/${configId}`);
  },

  analyzeDimensions: async (projectId: string, configId: string): Promise<DimensionAnalysis> => {
    const response = await apiClient.get<DimensionAnalysis>(
      `/projects/${projectId}/hierarchical-models/${configId}/dimensions`
    );
    return response.data;
  },

  startTraining: async (projectId: string, configId: string): Promise<{ status: string; task_id: string }> => {
    const response = await apiClient.post<{ status: string; task_id: string }>(
      `/projects/${projectId}/hierarchical-models/${configId}/train`
    );
    return response.data;
  },

  getStatus: async (projectId: string, configId: string): Promise<TrainingStatus> => {
    const response = await apiClient.get<TrainingStatus>(
      `/projects/${projectId}/hierarchical-models/${configId}/status`
    );
    return response.data;
  },

  cancelTraining: async (projectId: string, configId: string): Promise<{ status: string }> => {
    const response = await apiClient.post<{ status: string }>(
      `/projects/${projectId}/hierarchical-models/${configId}/cancel`
    );
    return response.data;
  },

  getResults: async (projectId: string, configId: string): Promise<ResultsSummary> => {
    const response = await apiClient.get<ResultsSummary>(
      `/projects/${projectId}/hierarchical-models/${configId}/results`
    );
    return response.data;
  },
};

File: backend/app/api/v1/endpoints/hierarchical.py
"""Hierarchical model endpoints."""

from uuid import UUID

from fastapi import APIRouter, HTTPException, status
from sqlalchemy import select
from sqlalchemy.orm import selectinload

from app.core.deps import CurrentUser, DbSession
from app.models import (
    Dataset,
    HierarchicalModelConfig,
    ModelConfig,
    Project,
    SubModelStatus,
)
from app.schemas.hierarchical import (
    DimensionAnalysis,
    DimensionCombination,
    HierarchicalModelCreate,
    HierarchicalModelList,
    HierarchicalModelRead,
    HierarchicalResultsSummary,
    HierarchicalTrainingStatus,
    SubModelRead,
    TrainingProgress,
)

router = APIRouter(prefix="/projects/{project_id}/hierarchical-models", tags=["Hierarchical Models"])


async def get_project_or_404(
    project_id: UUID,
    db: DbSession,
    current_user: CurrentUser,
) -> Project:
    """Get project or raise 404."""
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return project


@router.get("", response_model=list[HierarchicalModelList])
async def list_hierarchical_models(
    project_id: UUID,
    db: DbSession,
    current_user: CurrentUser,
):
    """List all hierarchical models in a project."""
    await get_project_or_404(project_id, db, current_user)

    result = await db.execute(
        select(HierarchicalModelConfig)
        .where(HierarchicalModelConfig.project_id == project_id)
        .options(selectinload(HierarchicalModelConfig.sub_models))
        .order_by(HierarchicalModelConfig.created_at.desc())
    )
    configs = result.scalars().all()

    return [
        HierarchicalModelList(
            id=c.id,
            name=c.name,
            granularity_type=c.granularity_type,
            model_type=c.model_type,
            status=c.status,
            sub_model_count=len(c.sub_models),
            completed_count=sum(1 for s in c.sub_models if s.status == SubModelStatus.COMPLETED.value),
            created_at=c.created_at,
        )
        for c in configs
    ]


@router.post("", response_model=HierarchicalModelRead, status_code=status.HTTP_201_CREATED)
async def create_hierarchical_model(
    project_id: UUID,
    data: HierarchicalModelCreate,
    db: DbSession,
    current_user: CurrentUser,
):
    """Create a new hierarchical model configuration."""
    await get_project_or_404(project_id, db, current_user)

    # Validate dataset exists
    result = await db.execute(
        select(Dataset).where(
            Dataset.id == data.dataset_id,
            Dataset.project_id == project_id,
        )
    )
    dataset = result.scalar_one_or_none()
    if not dataset:
        raise HTTPException(status_code=404, detail="Dataset not found")

    # Validate parent model if specified
    if data.parent_model_id:
        result = await db.execute(
            select(ModelConfig).where(
                ModelConfig.id == data.parent_model_id,
                ModelConfig.project_id == project_id,
            )
        )
        parent_model = result.scalar_one_or_none()
        if not parent_model:
            raise HTTPException(status_code=404, detail="Parent model not found")

    # Create hierarchical config
    config = HierarchicalModelConfig(
        project_id=project_id,
        name=data.name,
        parent_model_id=data.parent_model_id,
        dataset_id=data.dataset_id,
        dimension_columns=data.dimension_columns,
        granularity_type=data.granularity_type,
        model_type=data.model_type,
        target_variable=data.target_variable,
        date_column=data.date_column,
        features=data.features,
        inherit_constraints=data.inherit_constraints,
        constraint_relaxation=data.constraint_relaxation,
        inherit_priors=data.inherit_priors,
        prior_weight=data.prior_weight,
        min_observations=data.min_observations,
        status="pending",
    )

    db.add(config)
    await db.commit()
    await db.refresh(config)

    return config


@router.get("/{config_id}", response_model=HierarchicalModelRead)
async def get_hierarchical_model(
    project_id: UUID,
    config_id: UUID,
    db: DbSession,
    current_user: CurrentUser,
):
    """Get a hierarchical model configuration."""
    await get_project_or_404(project_id, db, current_user)

    result = await db.execute(
        select(HierarchicalModelConfig)
        .where(
            HierarchicalModelConfig.id == config_id,
            HierarchicalModelConfig.project_id == project_id,
        )
        .options(selectinload(HierarchicalModelConfig.sub_models))
    )
    config = result.scalar_one_or_none()

    if not config:
        raise HTTPException(status_code=404, detail="Hierarchical model not found")

    return config


@router.delete("/{config_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_hierarchical_model(
    project_id: UUID,
    config_id: UUID,
    db: DbSession,
    current_user: CurrentUser,
):
    """Delete a hierarchical model configuration."""
    await get_project_or_404(project_id, db, current_user)

    result = await db.execute(
        select(HierarchicalModelConfig).where(
            HierarchicalModelConfig.id == config_id,
            HierarchicalModelConfig.project_id == project_id,
        )
    )
    config = result.scalar_one_or_none()

    if not config:
        raise HTTPException(status_code=404, detail="Hierarchical model not found")

    await db.delete(config)
    await db.commit()


@router.get("/{config_id}/dimensions", response_model=DimensionAnalysis)
async def analyze_dimensions(
    project_id: UUID,
    config_id: UUID,
    db: DbSession,
    current_user: CurrentUser,
):
    """Analyze dimension combinations in the dataset."""
    import pandas as pd

    from app.services.file_storage import FileStorage

    await get_project_or_404(project_id, db, current_user)

    result = await db.execute(
        select(HierarchicalModelConfig)
        .where(
            HierarchicalModelConfig.id == config_id,
            HierarchicalModelConfig.project_id == project_id,
        )
        .options(selectinload(HierarchicalModelConfig.dataset))
    )
    config = result.scalar_one_or_none()

    if not config:
        raise HTTPException(status_code=404, detail="Hierarchical model not found")

    # Load dataset
    storage = FileStorage()
    df = pd.read_csv(storage.get_path(config.dataset.file_path))

    # Validate dimension columns exist
    for col in config.dimension_columns:
        if col not in df.columns:
            raise HTTPException(status_code=400, detail=f"Dimension column '{col}' not found in dataset")

    # Get unique combinations
    if len(config.dimension_columns) == 1:
        col = config.dimension_columns[0]
        combinations = [
            DimensionCombination(
                values={col: str(v)},
                observation_count=int((df[col] == v).sum()),
            )
            for v in df[col].unique()
        ]
    else:
        grouped = df.groupby(config.dimension_columns).size().reset_index(name="count")
        combinations = [
            DimensionCombination(
                values={col: str(row[col]) for col in config.dimension_columns},
                observation_count=int(row["count"]),
            )
            for _, row in grouped.iterrows()
        ]

    return DimensionAnalysis(
        dimension_columns=config.dimension_columns,
        combinations=combinations,
        total_combinations=len(combinations),
    )


@router.post("/{config_id}/train", status_code=status.HTTP_202_ACCEPTED)
async def start_training(
    project_id: UUID,
    config_id: UUID,
    db: DbSession,
    current_user: CurrentUser,
):
    """Start training all sub-models."""
    from app.workers.tasks import train_hierarchical_model

    await get_project_or_404(project_id, db, current_user)

    result = await db.execute(
        select(HierarchicalModelConfig).where(
            HierarchicalModelConfig.id == config_id,
            HierarchicalModelConfig.project_id == project_id,
        )
    )
    config = result.scalar_one_or_none()

    if not config:
        raise HTTPException(status_code=404, detail="Hierarchical model not found")

    if config.status == "training":
        raise HTTPException(status_code=400, detail="Training already in progress")

    # Start Celery task
    task = train_hierarchical_model.delay(str(config_id))

    # Update status
    config.status = "training"
    config.task_id = task.id
    await db.commit()

    return {
        "status": "started",
        "task_id": task.id,
        "config_id": str(config_id),
    }


@router.get("/{config_id}/status", response_model=HierarchicalTrainingStatus)
async def get_training_status(
    project_id: UUID,
    config_id: UUID,
    db: DbSession,
    current_user: CurrentUser,
):
    """Get training status for a hierarchical model."""
    await get_project_or_404(project_id, db, current_user)

    result = await db.execute(
        select(HierarchicalModelConfig)
        .where(
            HierarchicalModelConfig.id == config_id,
            HierarchicalModelConfig.project_id == project_id,
        )
        .options(selectinload(HierarchicalModelConfig.sub_models))
    )
    config = result.scalar_one_or_none()

    if not config:
        raise HTTPException(status_code=404, detail="Hierarchical model not found")

    # Calculate progress
    total = len(config.sub_models)
    completed = sum(1 for s in config.sub_models if s.status == SubModelStatus.COMPLETED.value)
    failed = sum(1 for s in config.sub_models if s.status == SubModelStatus.FAILED.value)
    in_progress = sum(1 for s in config.sub_models if s.status == SubModelStatus.TRAINING.value)
    pending = sum(1 for s in config.sub_models if s.status == SubModelStatus.PENDING.value)

    return HierarchicalTrainingStatus(
        status=config.status,
        progress=TrainingProgress(
            total=total,
            completed=completed,
            failed=failed,
            in_progress=in_progress,
            pending=pending,
        ),
        sub_models=[SubModelRead.model_validate(s) for s in config.sub_models],
    )


@router.post("/{config_id}/cancel", status_code=status.HTTP_200_OK)
async def cancel_training(
    project_id: UUID,
    config_id: UUID,
    db: DbSession,
    current_user: CurrentUser,
):
    """Cancel training for a hierarchical model."""
    from app.workers.celery_app import celery_app

    await get_project_or_404(project_id, db, current_user)

    result = await db.execute(
        select(HierarchicalModelConfig).where(
            HierarchicalModelConfig.id == config_id,
            HierarchicalModelConfig.project_id == project_id,
        )
    )
    config = result.scalar_one_or_none()

    if not config:
        raise HTTPException(status_code=404, detail="Hierarchical model not found")

    if config.task_id:
        # Revoke the task
        celery_app.control.revoke(config.task_id, terminate=True)

    config.status = "cancelled"
    await db.commit()

    return {"status": "cancelled"}


@router.get("/{config_id}/results", response_model=HierarchicalResultsSummary)
async def get_results_summary(
    project_id: UUID,
    config_id: UUID,
    db: DbSession,
    current_user: CurrentUser,
):
    """Get summary of hierarchical model results."""
    await get_project_or_404(project_id, db, current_user)

    result = await db.execute(
        select(HierarchicalModelConfig)
        .where(
            HierarchicalModelConfig.id == config_id,
            HierarchicalModelConfig.project_id == project_id,
        )
        .options(
            selectinload(HierarchicalModelConfig.sub_models),
            selectinload(HierarchicalModelConfig.parent_model),
        )
    )
    config = result.scalar_one_or_none()

    if not config:
        raise HTTPException(status_code=404, detail="Hierarchical model not found")

    completed_subs = [s for s in config.sub_models if s.status == SubModelStatus.COMPLETED.value]
    r_squared_values = [s.r_squared for s in completed_subs if s.r_squared is not None]

    # Build coefficient comparisons (simplified - would need to load full results)
    coefficient_comparisons = []

    return HierarchicalResultsSummary(
        id=config.id,
        name=config.name,
        total_sub_models=len(config.sub_models),
        completed_sub_models=len(completed_subs),
        avg_r_squared=sum(r_squared_values) / len(r_squared_values) if r_squared_values else None,
        min_r_squared=min(r_squared_values) if r_squared_values else None,
        max_r_squared=max(r_squared_values) if r_squared_values else None,
        coefficient_comparisons=coefficient_comparisons,
    )
